### PR TITLE
Make type error fatal to prevent assert failure at later point

### LINKED ISSUE #8283: ICE in solidity::frontend::TypeChecker::visit, Cannot declare variable with void (empty tuple) type.
## Description
Consider the following POC:
```
contract C {
    function f ( ) public {
        var i = ( ( 1 ( 3 ) + address ) , 2 ) ** 3 ;
    }
}
```

Run with `solc --bin -o poc poc.sol` and we got 
```
Internal compiler error during compilation:
/home/yongheng/solidity_clean/libsolidity/analysis/TypeChecker.cpp(1118): Throw in function virtual bool solidity::frontend::TypeChecker::visit(const solidity::frontend::VariableDeclarationStatement &)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Cannot declare variable with void (empty tuple) type.
[solidity::util::tag_comment*] = Cannot declare variable with void (empty tuple) type.
```
## Environment

- Compiler version: Version: 0.6.3-develop.2020.2.10+commit.d033c2f7.Linux.clang
- Target EVM version (as per compiler settings):
- Framework/IDE (e.g. Truffle or Remix):
- EVM execution environment / backend / blockchain client:
- Operating system: Ubuntu 16

### PR DESCRIPTION
**Pull request by Marenz**
fixes #8283, #8272

### CODE REVIEW

File: test/libsolidity/syntaxTests/emit/emit_non_event.sol [hunk 1]
```diff
@@ -2,9 +2,8 @@ contract C {
   uint256 Test;
 
   function f() public {
-    emit Test();
+    emit f();
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/emit/emit_non_event.sol]: This is not what the test is supposed to test...

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: test/libsolidity/syntaxTests/emit/emit_non_event.sol (hunk 1 above)
Comment (Marenz) [reply test/libsolidity/syntaxTests/emit/emit_non_event.sol]: I thought the idea was to test the code that produces the error message "must emit event" and that one is only output if you use `f()`, if you use `Test()` you _only_ get "`Test is not callable`"

Comment (ekpyron) [reply test/libsolidity/syntaxTests/emit/emit_non_event.sol]: Ah well, ok maybe it is, but then you can remove the ``uint256 Test;`` above.

Comment (ekpyron) [reply test/libsolidity/syntaxTests/emit/emit_non_event.sol]: Or maybe best: change ``Test`` to be something callable.

Comment (Marenz) [reply test/libsolidity/syntaxTests/emit/emit_non_event.sol]: updated

**Review by ekpyron — APPROVED** [review]
Yeah, I was wondering whether in this particular case, this should really be fatal, but the number of test cases we have about this with comments like "this used to crash" kind of confirms that it's a good idea to just make this fatal :-).

**Comment by ekpyron** [comment]
But maybe (independently of this PR) we should have a general discussion about whether to generally *always* make type errors fatal, whenever you cannot reasonably assign types to all expressions and document that somewhere...

Especially, since in the long term and considering stuff like error recovery we might want to deal with all these cases more gracefully...

**Comment by Marenz** [comment]
Alternatively we could adopt something like an "ErrorType" used in places where the code was supposed to assign a type but errored (and similar dummy constructs)

**Comment by ekpyron** [comment]
Yeah... the problem with that is that we'd probably need special cases for the error type all over... at least that *could* happen and wouldn't be nice...

**Comment by chriseth** [comment]
This is missing a changelog entry.

**Comment by chriseth** [comment]
Created https://github.com/ethereum/solidity/issues/8316 so that we do not forget it.

**Pull request merged**